include/unifex/via.hpp

```diff
@@ -30,9 +30,229 @@
+    unifex::set_error(std::forward<Receiver>(receiver_), std::move(error_));
+  }
+
+  template <typename OtherError>
+  void set_error(OtherError&& otherError) noexcept {
+    unifex::set_error(
+        std::forward<Receiver>(receiver_), (OtherError &&) otherError);
+  }
+
+  void set_done() noexcept {
+    unifex::set_done(std::forward<Receiver>(receiver_));
+  }
+
+  template <
+      typename CPO,
+      std::enable_if_t<!is_receiver_cpo_v<CPO>, int> = 0>
+  friend auto tag_invoke(CPO cpo, const error_receiver& r) noexcept(
+      std::is_nothrow_invocable_v<CPO, const Receiver&>)
+      -> std::invoke_result_t<CPO, const Receiver&> {
+    return std::move(cpo)(std::as_const(r.receiver_));
+  }
+
+  template <typename Func>
+  friend void tag_invoke(
+      tag_t<visit_continuations>,
+      const error_receiver& r,
+      Func&& func) {
+    std::invoke(func, r.receiver_);
+  }
+};
+
+template <typename Receiver>
+struct done_receiver_ {
+  struct type;
+};
+template <typename Receiver>
+using done_receiver = typename done_receiver_<Receiver>::type;
+
+template <typename Receiver>
+struct done_receiver_<Receiver>::type {
+  using done_receiver = type;
+  UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+
+  void set_value() noexcept {
+    unifex::set_done(std::forward<Receiver>(receiver_));
+  }
+
+  template <typename OtherError>
+  void set_error(OtherError&& otherError) noexcept {
+    unifex::set_error(
+        std::forward<Receiver>(receiver_), (OtherError &&) otherError);
+  }
+
+  void set_done() noexcept {
+    unifex::set_done(std::forward<Receiver>(receiver_));
+  }
+
+  template <
+      typename CPO,
+      std::enable_if_t<!is_receiver_cpo_v<CPO>, int> = 0>
+  friend auto tag_invoke(CPO cpo, const done_receiver& r) noexcept(
+      std::is_nothrow_invocable_v<CPO, const Receiver&>)
+      -> std::invoke_result_t<CPO, const Receiver&> {
+    return std::move(cpo)(std::as_const(r.receiver_));
+  }
+
+  template <typename Func>
+  friend void tag_invoke(
+      tag_t<visit_continuations>,
+      const done_receiver& r,
+      Func&& func) {
+    std::invoke(func, r.receiver_);
+  }
+};
+
+template <typename Successor, typename Receiver>
+struct predecessor_receiver_ {
+  struct type;
+};
+template <typename Successor, typename Receiver>
+using predecessor_receiver =
+    typename predecessor_receiver_<Successor, std::remove_cvref_t<Receiver>>::type;
+
+template <typename Successor, typename Receiver>
+struct predecessor_receiver_<Successor, Receiver>::type {
+  using predecessor_receiver = type;
+  Successor successor_;
+  Receiver receiver_;
+
+  template <typename... Values>
+  void set_value(Values&&... values) && noexcept {
+    try {
+      submit(
+          (Successor &&) successor_,
+          value_receiver<Receiver, Values...>{
+              {(Values &&) values...}, (Receiver &&) receiver_});
+    } catch (...) {
+      unifex::set_error(
+          static_cast<Receiver&&>(receiver_), std::current_exception());
+    }
```

Comment:
I've just noticed that this implementation suffers from the same problem that we found in the P0443R13 spec. If an exception is thrown then the receiver may be in a moved-from state and it may not be valid to call set_error().

The plan is to remove this via() and replace it with typed_via() so it may just sort itself out.